### PR TITLE
ci: schema json generation ci check

### PR DESCRIPTION
**Pull request by james-milligan**
Signed-off-by: James Milligan <james@omnant.co.uk>

## This PR

- adds `gen-schema-json` makefile command to generate json definition from yaml schema
- adds ci step to ensure there is no diff when `gen-schema-json` command is run

### Follow-up Tasks

Open new issue to expand `CONTRIBUTING.md` with guidance for contributing to the protobuf definitions

### CODE REVIEW

File: json/flagd-definitions.yaml [hunk 1]
```diff
@@ -1,39 +1,36 @@
...
   flag:
     title: Flag base
     description: Base object for all flags
     properties:
       state:
-        "title:": Flag state
-        description:
-          Indicates whether the flag is functional. Disabled flags are
-          treated as if they don't exist
+        'title:': Flag state
```
Comment (beeme1mr) [inline json/flagd-definitions.yaml:32]:
```suggestion
        'title:': Flag State
```

File: json/flagd-definitions.yaml [hunk 2]
```diff
@@ -1,39 +1,36 @@
...
+            $ref: '#/$defs/numberFlag'
           - title: Object flag
             description: A flag associated with arbitrary object values
-            "$ref": "#/$defs/objectFlag"
+            $ref: '#/$defs/objectFlag'
         unevaluatedProperties: false
-"$defs":
+$defs:
   flag:
     title: Flag base
```
Comment (beeme1mr) [inline json/flagd-definitions.yaml:28]:
```suggestion
    title: Flag Base
```

File: json/flagd-definitions.yaml [hunk 3]
```diff
@@ -1,39 +1,36 @@
----
-"$schema": http://json-schema.org/draft-07/schema#
+$schema: http://json-schema.org/draft-07/schema#
 title: flagd Flag Configuration
```
Comment (beeme1mr) [inline json/flagd-definitions.yaml:2]:
```suggestion
title: Flagd Flag Configuration
```

Comment (toddbaert) [reply json/flagd-definitions.yaml]: no

File: json/flagd-definitions.yaml [hunk 4]
```diff
@@ -45,10 +42,7 @@ properties:
       targeting:
         type: object
         title: Targeting Logic
-        description:
-          JsonLogic expressions to be used for dynamic evaluation. The
-          "context" is passed as the data. Rules must resolve one of the defined variants,
-          or the "defaultVariant" will be used.
+        description: JsonLogic expressions to be used for dynamic evaluation. The "context" is passed as the data. Rules must resolve one of the defined variants, or the "defaultVariant" will be used.
```
Comment (toddbaert) [inline json/flagd-definitions.yaml:45]:
```suggestion
        description: JsonLogic expressions to be used for dynamic evaluation. The "evaluation context" is passed as the data. Rules must resolve one of the defined variants, or the "defaultVariant" will be used.
```

**Review by toddbaert — APPROVED** [review]

File: Makefile [hunk 5]
```diff
@@ -40,4 +40,4 @@ gen-ruby: install-buf guard-GOPATH
 	${GOPATH}/bin/buf generate buf.build/open-feature/flagd --template protobuf/buf.gen.ruby.yaml
 
 gen-schema-yaml: install-yq
-	yq eval -j json/flagd-definitions.json | yq -P > json/flagd-definitions.yaml
+	yq eval -o=json json/flagd-definitions.json | yq -P > json/flagd-definitions.yaml
```
Comment (beeme1mr) [inline Makefile:43]: I wonder if we should consider going the other way. Manually managing a yaml file tends to be easier than json. What do you think?

Comment (james-milligan) [reply Makefile:43]: Im happy to make this change, I agree it'll be easier to manage

File: .github/workflows/pr-checks.yml [hunk 6]
```diff
@@ -63,3 +63,9 @@ jobs:
       - name: Validate Schema
         run: go test -v ./json
 
+      # Ensure there is no dif when make gen-schema-json is run
```
Comment (skyerus) [inline .github/workflows/pr-checks.yml:66]:
```suggestion
      # Ensure there is no diff when make gen-schema-json is run 
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
+## Guidelines
+
+Welcome!
+
+Please ensure you have read the [a code of conduct](https://github.com/open-feature/.github/blob/main/CODE_OF_CONDUCT.md) before contributing.
```
Comment (skyerus) [inline CONTRIBUTING.md:5]:
```suggestion
Please ensure you have read the [code of conduct](https://github.com/open-feature/.github/blob/main/CODE_OF_CONDUCT.md) before contributing.  
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+Please ensure you have read the [a code of conduct](https://github.com/open-feature/.github/blob/main/CODE_OF_CONDUCT.md) before contributing.  
+TLDR: be respectful.
+
+
+## Developing
+
+### flagd JSON schema
+
+The flagd JSON schema defines the json structure used within flagd for flag configurations.   
+To contribute to the schema make changes in `json/flagd-definitions.yaml`, these changes can then be reflected in `json/flagd-definitions.json` by running `make gen-schema-json`.
```
Comment (skyerus) [inline CONTRIBUTING.md:14]:
```suggestion
To contribute to the schema, make changes in `json/flagd-definitions.yaml`, these changes should then be propagated to `json/flagd-definitions.json` by running `make gen-schema-json`.  
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+TLDR: be respectful.
+
+
+## Developing
+
+### flagd JSON schema
+
+The flagd JSON schema defines the json structure used within flagd for flag configurations.   
+To contribute to the schema make changes in `json/flagd-definitions.yaml`, these changes can then be reflected in `json/flagd-definitions.json` by running `make gen-schema-json`.  
+Unit testing for the schema is managed through the `json/test/negative` (in which all json files must fail validation against the schema) and `json/test/positive` (in which all json files must pass).
```
Comment (skyerus) [inline CONTRIBUTING.md:15]:
```suggestion
Unit testing for the schema is managed through [json/test/negative](./json/test/negative) (in which all json files must fail validation against the schema) and [json/test/positive](./json/test/positive) (in which all json files must pass).
```

File: json/flagd-definitions.yaml (hunk 3 above)
Comment (beeme1mr) [reply json/flagd-definitions.yaml:2]: Even in a title? We need a naming guide so we can be consistent.

**Review by toddbaert — APPROVED** [review]

File: json/flagd-definitions.yaml (hunk 3 above)
Comment (toddbaert) [reply json/flagd-definitions.yaml:2]: I'd be happy to write one.

But ya, IMHO, stylized names like "flagd" should not be capitalized, even at the start of a sentence. See etcd docs: https://etcd.io/docs/v3.5/learning/design-client/#introduction

**Pull request merged**